includes/admin/class-wc-admin-post-types.php

```diff
@@ -986,6 +868,103 @@ public function add_display_post_states( $post_states, $post ) {
 
 		return $post_states;
 	}
+
+	/**
+	 * Apply product type constraints to stock status.
+	 *
+	 * @param WC_Product $product The product whose stock status will be adjusted.
+	 * @param string     $stock_status The stock status to use for adjustment.
+	 * @return WC_Product The supplied product, or the synced product if it was a variable product.
+	 */
+	private function update_stock_status( $product, $stock_status ) {
+		if ( $product->is_type( 'external' ) ) {
+			// External products are always in stock.
+			$product->set_stock_status( 'instock' );
+		} elseif ( $product->is_type( 'variable' ) && ! $product->get_manage_stock() ) {
+			// Stock status is determined by children.
+			foreach ( $product->get_children() as $child_id ) {
+				$child = wc_get_product( $child_id );
+				if ( ! $product->get_manage_stock() ) {
+					$child->set_stock_status( $stock_status );
+					$child->save();
+				}
+			}
+			$product = WC_Product_Variable::sync( $product, false );
+		} else {
+			$product->set_stock_status( $stock_status );
+		}
+
+		return $product;
+	}
+
+	/**
+	 * Set the new regular or sale price if requested.
+	 *
+	 * @param WC_Product $product The product to set the new price for.
+	 * @param string     $price_type 'regular' or 'sale'.
+	 * @return bool true if a new price has been set, false otherwise.
+	 */
+	private function set_new_price( $product, $price_type ) {
+		// phpcs:disable WordPress.Security.NonceVerification.Recommended
+
+		if ( empty( $_REQUEST[ "change_{$price_type}_price" ] ) || ! isset( $_REQUEST[ "_{$price_type}_price" ] ) ) {
```

Comment:
Just came across this commit, not a review and you probably already plan to do this, but since you are refactoring, it's probably better to shift `_REQUEST` calls in parent methods and pass the values in variables. This way method can be used outside of a `_REQUEST` context and unit testing will be slightly easier :).

Reply:
Hummm... yes and no. 🙂

I agree on extracting/wrapping `_REQUEST` somehow instead of using it directly, to ease unit testing. I also agree that methods should generally receive only the data they need. However in this particular case I prefer to access the request data directly instead of having it passed as parameters, otherwise I would have to extract the initial `if ( empty( ...` line out and that would add some duplication.

If at some point we really need to use this method outside of the context of a request, then sure, we can give it another shot and think how to make it completely request-free.

That said, I have created a new `request_data` method that just returns `_$REQUEST`, and changed the entire class to use it instead of the superglobal. Eventually we should find a way to do it consistently throughout the entire codebase, e.g. making it a method in an abstract base class.